packages/codec/lib/abi-data/allocate/index.ts

```diff
@@ -427,33 +452,73 @@ function allocateCalldata(
     //if node doesn't exist, OR if something went wrong
     //during allocation
     id = "-1"; //fake irrelevant ID
-    parameterTypes = abiEntry.inputs.map(parameter => ({
+    inputTypes = abiEntry.inputs.map(parameter => ({
       name: parameter.name,
       type: Import.abiParameterToType(parameter)
     }));
-    abiAllocation = allocateMembers(
+    if (abiEntry.type === "function") {
+      outputTypes = abiEntry.outputs.map(parameter => ({
+        name: parameter.name,
+        type: Import.abiParameterToType(parameter)
+      }));
+    } else {
+      //let's just set it empty for now, we can set
+      //the bytecode thingy later
+      outputTypes = [];
+    }
+    abiAllocationInput = allocateMembers(
       id,
-      parameterTypes,
+      inputTypes,
       userDefinedTypes,
       abiAllocations,
       offset
     )[id];
+    abiAllocationOutput = allocateMembers(
+      id,
+      outputTypes,
+      userDefinedTypes,
+      abiAllocations
+      //note the lack of an offset!
+    )[id];
   }
   //finally: transform the allocation appropriately
-  let argumentsAllocation = abiAllocation.members.map(member => ({
+  let inputArgumentsAllocation = abiAllocationInput.members.map(member => ({
     ...member,
     pointer: {
       location: "calldata" as const,
       start: member.pointer.start,
       length: member.pointer.length
     }
   }));
-  return {
+  let outputArgumentsAllocation = abiAllocationOutput.members.map(member => ({
+    ...member,
+    pointer: {
+      location: "returndata" as const,
+      start: member.pointer.start,
+      length: member.pointer.length
+    }
+  }));
+  let inputsAllocation = {
     abi: abiEntry,
     offset,
-    arguments: argumentsAllocation,
+    arguments: inputArgumentsAllocation,
     allocationMode
   };
+  let outputsAllocation;
+  switch (abiEntry.type) {
+    case "function":
+      outputsAllocation = {
+        selector: new Uint8Array(), //empty by default
+        arguments: outputArgumentsAllocation,
+        allocationMode,
+        kind: "return" as const
+      };
+      break;
+    case "constructor":
+      outputsAllocation = CONSTRUCTOR_OUTPUT_ALLOCATION;
+      break;
+  }
+  return { input: inputsAllocation, output: outputsAllocation };
```

Comment:
(Also, more generally beyond just this function: I'm curious to see what this module would look like if it were factored into common/calldata-specific/return-specific)

Reply:
Hm, I guess this could be factored, yeah.  Also spotting some outdated comments in there...

Reply:
Oh wait -- here's a reason it's not factored like that.  Getting the parameters!

I wrote `getterParameters` to get both input and output parameters.  Notionally this could be two separate functions, of course, but I wrote it as one.  So, one reason I didn't factor it like you suggest was that that would involve calling the function twice.

Perhaps this sort of optimization is silly, but I like it how it is.  Thoughts?

Reply:
I also like the optimization; can't you preserve that in factoring? Just split off from the point of that dual-return, passing inputs/outputs individually to a factored function for handling either

Reply:
Oh, yeah, that would work.

Reply:
Blech, turns out that doesn't work so simply after all, because it doesn't really work very well with the whole full-mode-abi-mode structure.  You think it's OK to leave this one as-is? :-/

Reply:
What's the problem with that?

I'm more strongly in favor of doing cleanup here explicitly because returndata is so wonky ("the wild west", as you called it :)

Reply:
OK.  So, let's consider this.

Right now I have the same mode being used for both input and output, and, um, actually I should likely change that.  But!  First let's discuss how things work at the moment.

First we find the node, if it exists.  If so, we set mode to full and get the full-mode parameters.  If not, we set mode to ABI.  Then, if mode is full, we attempt to allocate based on the full-mode parameters.  If this fails, we set mode to ABI.  Then, if mode is ABI (whether because we failed to find the node or because allocation failed), we get the ABI-mode parameters, and allocate based on that.

So, one reason this is hard to factor at the moment is because I used the same mode for both.  But likely I shouldn't do that, and should change that regardless of whether I try to factor like you say.  So let's say I do that, which is easy.  Factoring still remains difficult unfortunately.  Let's see why.

OK.  So say we now have two separate modes.  How does the factored version work?  It will find the node (if it exists).  It will get parameters, since we want to do that for both input and output simultaneously.  It'll have to get both full-mode and ABI-mode parameters, regardless of mode, since it'll need to pass both of these to the new function it's calling; the new function it's calling can't just get the ABI parameters if something goes wrong.  Well, getting ABI-mode parameters isn't a problem, so that's fine, right?

OK, but what if there was no node?  Then we can't get the full-mode parameters at all.  I guess we could pass undefined, or something?  Either that or we'd have to pass in whether we're starting out in full mode or ABI mode!

I guess the first way is viable, with passing undefined, but at this point the new function being called is doing so little I'm not sure it's worth it...

Reply:
Actually, rethinking it again, I think factoring it makes sense after all.  Because I was thinking, what would it actually look like if I made two separate modes?  And the result was something that should clearly be factored.  So, I'll do that.

Reply:
OK, factored this and made the two modes separate.

Reply:
Note that I maybe didn't factor this as much as you would've liked, but more would've been messy.